File: samples/Clojure/zebra-puzzle.clj
(ns zebra.core
  (:refer-clojure :exclude [==])
  (:use [clojure.core.logic]
        [clojure.tools.macro :as macro]))

(defne lefto [x y l]
       ([_ _ [x y . ?r]])
       ([_ _ [_ . ?r]] (lefto x y ?r)))

(defn nexto [x y l]
  (conde
    ((lefto x y l))
    ((lefto y x l))))

(defn zebrao [hs]
  (macro/symbol-macrolet [_ (lvar)]
    (all
      (== [_ _ _ _ _] hs)
      (membero ['englishman _ _ _ 'red] hs)
      (membero ['swede _ _ 'dog _] hs)
      (membero ['dane _ 'tea _ _] hs)
      (lefto [_ _ _ _ 'green] [_ _ _ _ 'white] hs)
      (membero [_ _ 'coffee _ 'green] hs)
      (membero [_ 'pallmall _ 'birds _] hs)
      (membero [_ 'dunhill _ _ 'yellow] hs)
      (== [_ _ [_ _ 'milk _ _] _ _ ] hs)
      (firsto hs ['norwegian _ _ _ _])
      (nexto [_ 'blend _ _ _] [_ _ _ 'cats _ ] hs)
      (nexto [_ _ _ 'horse _] [_ 'dunhill _ _ _] hs)
      (membero [_ 'bluemaster 'beer _ _] hs)
      (membero ['german 'prince _ _ _] hs)
      (nexto ['norwegian _ _ _ _] [_ _ _ _ 'blue] hs)
      (nexto [_ _ 'water _ _] [_ 'blend _ _ _] hs)
      (membero [_ _ _ 'zebra _] hs))))

(let [solns (run* [q] (zebrao q))
      soln (first solns)
      zebra-owner (->> soln (filter #(= 'zebra (% 3))) first (#(% 0)))]
  (println "solution count:" (count solns))
  (println "zebra owner is the" zebra-owner)
  (println "full solution (in house order):")
  (doseq [h soln] (println " " h)))

File: samples/Haskell/zebra-puzzle.hs
import Control.Applicative ((<$>), (<*>))
import Control.Monad
import Data.List ((\\), isInfixOf)

-- types
data House = House
    { color :: Color
    , man   :: Man
    , pet   :: Pet
    , drink :: Drink
    , smoke :: Smoke
    }
    deriving (Eq, Show)

data Color = Red | Green | Blue | Yellow | White
    deriving (Eq, Show, Enum, Bounded)

data Man = Eng | Swe | Dan | Nor | Ger
    deriving (Eq, Show, Enum, Bounded)

data Pet = Dog | Birds | Cats | Horse | Zebra
    deriving (Eq, Show, Enum, Bounded)

data Drink = Coffee | Tea | Milk | Beer | Water
    deriving (Eq, Show, Enum, Bounded)

data Smoke = PallMall | Dunhill | Blend | BlueMaster | Prince
    deriving (Eq, Show, Enum, Bounded)


main :: IO ()
main = do
  mapM_ (\x-> mapM_ print (reverse x) >> putStrLn "----") solutions
  putStrLn "No More Solutions"


solutions :: [[House]]
solutions = filter (and . postChecks) $ foldM next [] [1..5]
    where
      next xs pos = [x:xs | x <- iterHouse xs, and $ checks pos x]


iterHouse :: [House] -> [House]
iterHouse xs =
    House <$> new color <*> new man <*> new pet <*> new drink <*> new smoke
    where
      new getter = [minBound ..] \\ map getter xs


-- immediate checks
checks :: Int -> House -> [Bool]
checks pos house =
    [ man   `is` Eng    <=> color `is` Red              --  2
    , man   `is` Swe    <=> pet   `is` Dog              --  3
    , man   `is` Dan    <=> drink `is` Tea              --  4
    , color `is` Green  <=> drink `is` Coffee           --  6
    , pet   `is` Birds  <=> smoke `is` PallMall         --  7
    , color `is` Yellow <=> smoke `is` Dunhill          --  8
    , const (pos == 3)  <=> drink `is` Milk             --  9
    , const (pos == 1)  <=> man   `is` Nor              -- 10
    , drink `is` Beer   <=> smoke `is` BlueMaster       -- 13
    , man   `is` Ger    <=> smoke `is` Prince           -- 14
    ]
    where
      infix 4 <=>
      p <=> q = p house == q house  -- both True or both False


-- final checks
postChecks :: [House] -> [Bool]
postChecks houses =
    -- NOTE: list of houses is generated in reversed order
    [ [White, Green] `isInfixOf` map color houses       --  5
    , (smoke `is` Blend  ) `nextTo` (pet   `is` Cats )  -- 11
    , (smoke `is` Dunhill) `nextTo` (pet   `is` Horse)  -- 12
    , (color `is` Blue   ) `nextTo` (man   `is` Nor  )  -- 15
    , (smoke `is` Blend  ) `nextTo` (drink `is` Water)  -- 16
    ]
    where
      nextTo :: (House -> Bool) -> (House -> Bool) -> Bool
      nextTo p q
          | (_:x:_) <- dropWhile (not . match) houses = match x
          | otherwise                                 = False
          where
            match x = p x || q x


is :: Eq a => (House -> a) -> a -> House -> Bool
getter `is` value = (== value) . getter
